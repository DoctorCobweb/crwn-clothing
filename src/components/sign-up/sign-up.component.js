import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import SweetAlert from 'sweetalert2-react';

import FormInput from '../form-input/form-input.component';
import CustomButton from '../custom-button/custom-button.component';

import {
  signUpStart
} from '../../redux/user/user.actions';

import { selectUserError } from '../../redux/user/user.selector';

import {
  SignUpContainer,
  SignUpTitle 
} from './sign-up.styles.jsx';

const SignUp = (props) => {
  const { signUpStart } = props
  const [ userCredentials, setUserCredentials ] = useState({
    displayName: '',
    email: '',
    password: '',
    confirmPassword: '',
    error: '',
  });

  const { displayName, email, password, confirmPassword, error } = userCredentials;


  // here is how useEffect can be made to work like componentWillReceiveProps lifecycle method.
  // but here we're specifically only listening for the userError prop which get sets
  // when theres an error during signup process
  useEffect(() => {
    if (props.userError) {
      setUserCredentials({ ...userCredentials, error: props.userError.message });
    }
  }, [props.userError]);

  const handleSubmit = async (event) => {
    event.preventDefault();
    if (password !== confirmPassword) {
      alert("passwords don't match");
      return;
    }
    signUpStart({ displayName, email, password })
  }

  const handleChange = event => {
    const { name, value } = event.target;
    setUserCredentials({ ...userCredentials, [name]: value });
  }

  return (
    <SignUpContainer>
      <SignUpTitle>I do not have an account</SignUpTitle>
      <span>Sign up with your email and password</span>
      <form className="sign-up-form" onSubmit={handleSubmit}>
        <FormInput
          type="text"
          name="displayName"
          value={displayName}
          onChange={handleChange}
          label='Display Name'
          required
        />
        <FormInput
          type="email"
          name="email"
          value={email}
          onChange={handleChange}
          label="Email"
          required
        />
        <FormInput
          type="password"
          name="password"
          value={password}
          onChange={handleChange}
          label="Password"
          required
        />
        <FormInput
          type="password"
          name="confirmPassword"
          value={confirmPassword}
          onChange={handleChange}
          label="Confirm Password"
          required
        />
        <CustomButton type="submit">SIGN UP</CustomButton>
        {
          error ? (
            <SweetAlert
              show={error}
              type="warning"
              title="Error"
              text={error}
              onConfirm={() => setUserCredentials({ ...userCredentials, error: ''})}
            />
          ) : (
            null
          )
        }
      </form>
    </SignUpContainer>
  );
}

const mapStateToProps = createStructuredSelector({
  userError: selectUserError,
})

const mapDispatchToProps = dispatch => ({
  signUpStart: (userCredentials) => dispatch(signUpStart(userCredentials)),
});

export default connect(mapStateToProps, mapDispatchToProps)(SignUp);