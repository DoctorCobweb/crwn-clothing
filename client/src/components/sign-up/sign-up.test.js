import React from 'react';
import { shallow } from 'enzyme';

import { SignUp } from './sign-up.component';
import FormInput from '../form-input/form-input.component';

describe('SignUp component', () => {
  let wrapper;
  let mockSignUpStart;
  const mockDisplayName = 'Blah';
  const mockEmail = 'blah@blah.com';
  const mockPassword = '1234';
  const mockPasswordConfirm = '1234';


  beforeEach(() => {
    mockSignUpStart = jest.fn(userCredentials => console.log(userCredentials));

    const mockProps = {
      signUpStart: mockSignUpStart,
    }

    wrapper = shallow(<SignUp {...mockProps} />);
  });

  it('should render SignUp correctly', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });


  it('should correctly set form inputs, via setState hook', () => {
    wrapper.find(FormInput).at(0).prop('onChange')({
      target: {
        value: mockDisplayName,
        name: 'displayName',
      }
    });
    wrapper.find(FormInput).at(1).prop('onChange')({
      target: {
        value: mockEmail,
        name: 'email',
      }
    });
    wrapper.find(FormInput).at(2).prop('onChange')({
      target: {
        value: mockPassword,
        name: 'password',
      }
    });
    wrapper.find(FormInput).at(3).prop('onChange')({
      target: {
        value: mockPasswordConfirm,
        name: 'confirmPassword',
      }
    });

    // make a fake event
    const fakeEvent = { preventDefault: () => {} };

    // submit the email sign in form
    wrapper.find('SignUpContainer').childAt(2).simulate('submit', fakeEvent);
    expect(mockSignUpStart).toHaveBeenCalled();
    expect(mockSignUpStart).toHaveBeenCalledWith({
      displayName: mockDisplayName,
      email: mockEmail,
      password: mockPassword,
    });
  });
});