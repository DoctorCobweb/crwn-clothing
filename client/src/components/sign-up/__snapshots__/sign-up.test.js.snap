// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp component should render SignUp correctly 1`] = `
"<SignUpContainer>
  <SignUpTitle>
    I do not have an account
  </SignUpTitle>
  <span>
    Sign up with your email and password
  </span>
  <form className=\\"sign-up-form\\" onSubmit={[Function: handleSubmit]}>
    <FormInput type=\\"text\\" name=\\"displayName\\" value=\\"\\" onChange={[Function: handleChange]} label=\\"Display Name\\" required={true} />
    <FormInput type=\\"email\\" name=\\"email\\" value=\\"\\" onChange={[Function: handleChange]} label=\\"Email\\" required={true} />
    <FormInput type=\\"password\\" name=\\"password\\" value=\\"\\" onChange={[Function: handleChange]} label=\\"Password\\" required={true} />
    <FormInput type=\\"password\\" name=\\"confirmPassword\\" value=\\"\\" onChange={[Function: handleChange]} label=\\"Confirm Password\\" required={true} />
    <CustomButton type=\\"submit\\">
      SIGN UP
    </CustomButton>
  </form>
</SignUpContainer>"
`;
